### PR TITLE
Add bhyve support to virt

### PR DESCRIPTION
**Pull request by jeroen92**
### What does this PR do?
Add bhyve support to the virt module, and add ZFS compatibility for cloning/creating zvols when using bhyve as a hypervisor.

### What issues does this PR fix or reference?
#47619

### Previous Behavior
Not implemented. Crashes when performing virt.init for a bhyve hypervisor.

### New Behavior
Using virt.init for a bhyve hypervisor will create a VM with (multiple) ZFS provisioned disks. Using virt.purge will destroy the VM and the corresponding ZFS datasets.

### Tests written?
No

### Commits signed with GPG?
No

### CODE REVIEW

File: salt/modules/virt.py [hunk 1]
```diff
@@ -145,6 +146,7 @@ def __esxi_auth():
                                     __esxi_auth(),
                                     0]],
         'qemu': [libvirt.open, [conn_str]],
+        'bhyve': [libvirt.open, ['bhyve:///system']],
```
Comment (cbosdo) [inline salt/modules/virt.py]: The whole connection mechanism has been rewritten recently. You don't need to add that anymore. See commit 2dfdc4eb4bdf9b168df53ee3416ebeb384d2247e for more details.

File: salt/modules/virt.py [hunk 2]
```diff
@@ -253,10 +255,7 @@ def _gen_xml(name,
         # TODO: make bus and model parameterized, this works for 64-bit Linux
         context['controller_model'] = 'lsilogic'
 
-    if kwargs.get('enable_vnc', True):
-        context['enable_vnc'] = True
-    else:
-        context['enable_vnc'] = False
+    context['enable_vnc'] = kwargs.get('enable_vnc', False)
```
Comment (cbosdo) [inline salt/modules/virt.py]: I know it's not yours, but I don't really like the `enable_vnc` option in the first place since that rules out the other graphic types supported by libvirt like spice. Refactoring it for a broader support would be good (I may get to it sooner or later)

File: salt/modules/virt.py [hunk 3]
```diff
@@ -273,21 +274,28 @@ def _gen_xml(name,
         else:
             context['telnet_port'] = 23023  # FIXME: use random unused port
     if 'serial_type' in context:
+        if context['serial_type'] == 'nmdm' and 'nmdm_id' in kwargs:
+            context['nmdm_id'] = kwargs['nmdm_id']
```
Comment (cbosdo) [inline salt/modules/virt.py]: It seems to me that this function is getting a lot of kwargs now, most of them undocumented. I wonder if we could come up with something cleaner than just adding kwargs every now and then.

**Review by cbosdo — CHANGES_REQUESTED** [review]
Your last two commits should be merged into one since they fix the first one. See my inline comments for the possible improvements, the most important one being rebasing on develop branch to benefit from the recent connection changes.

File: salt/modules/virt.py [hunk 4]
```diff
@@ -1741,6 +1875,19 @@ def is_xen_hyper():
     return 'libvirtd' in __salt__['cmd.run'](__grains__['ps'])
 
 
+def is_bhyve_hyper():
```
Comment (cbosdo) [inline salt/modules/virt.py]: For more flexibility I would deprecate the `is_.*hyper` functions and replace them by a `hypervisor_type` one.

**Comment by cbosdo** [comment]
I'm just starting to dive into the profiles system to improve it... and the more I think about it, the more I feel the hypervisor should be deduced from the libvirt connection that is used. This would make things simpler and more reliable in the virt module

**Comment by nbraud** [comment]
@jeroen92 Ping?

I might have a stab at rebasing this and pushing it past the finish line, if you are too busy/unavailable.

File: salt/modules/virt.py (hunk 2 above)
Comment (astronouth7303) [reply salt/modules/virt.py]: This changes the default value of `enable_vnc`. I'm not sure if you're ok with that.

**Comment by jeroen92** [comment]
Pong :)

I just got back from a holiday. I'll take a look at the comments and update the MR accordingly.

**Comment by cbosdo** [comment]
@jeroen92 as I said to @nbraud yesterday on IRC, there is a massive refactoring of the virt module on going: https://github.com/cbosdo/salt/tree/virt-init It would be great to rebase your work on this once it stabilizes a bit (in a week or so) as I'm still working on it.

**Comment by nbraud** [comment]
@jeroen92 OK, thanks a lot  :)

**Comment by rallytime** [comment]
@jeroen92 and @cbosdo where did we land with this change? I think a lot of the work @cbosdo has been working was merged in recently. (And this PR will need to be rebased once it's time to work on this again as there is a merge conflict.

Let us know when this is ready. :)

**Comment by jeroen92** [comment]
@rallytime Ack, didn't know @cbosdo his changes were merged in already. Although I'm a bit short on time this week, I think I should be able to fix up this MR.

**Comment by jeroen92** [comment]
@rallytime @cbosdo This MR has been updated :)

**Comment by rallytime** [comment]
@jeroen92 Looks like there are some minor lint failures on this change: https://jenkinsci.saltstack.com/job/pr-lint/job/PR-47620/9/warnings52Result/new/

And one related test is failing: `unit.modules.test_virt.VirtTestCase.test_capabilities`

https://jenkinsci.saltstack.com/job/pr-kitchen-ubuntu1604-py2/job/PR-47620/2/

Mind fixing those up in the mean time?

File: salt/modules/virt.py [hunk 5]
```diff
@@ -463,6 +464,7 @@ def _get_disks(dom):
     Get domain disks from a libvirt domain object.
     '''
     disks = {}
+    hypervisor = __salt__['config.get']('libvirt:hypervisor', 'kvm')
```
Comment (cbosdo) [inline salt/modules/virt.py]: don't use the ``libvirt:hypervisor`` configuration: you should rather use the domain type. I have a pending commit here that removes the use of minidom though the whole module, but you can temporarily do it with minidom and I'll rebase my change later on.

Note that ``libvirt:hypervisor`` is now deprecated.

File: salt/modules/virt.py [hunk 6]
```diff
@@ -600,6 +607,8 @@ def _gen_xml(name,
     else:
         context['boot_dev'] = ['hd']
 
+    if 'loader' in kwargs:
+        context['loader'] = kwargs['loader']
```
Comment (cbosdo) [inline salt/modules/virt.py]: There are enough ``**kwargs`` to be removed in that module, don't add yet another one. Add a normal parameter with default value instead.

File: salt/modules/virt.py (hunk 3 above)
Comment (cbosdo) [reply salt/modules/virt.py]: That comment still applies: ``nmdm_id`` should be moved into a normal parameter with default value rather than a ``**kwargs``.

File: salt/modules/virt.py [hunk 7]
```diff
@@ -750,6 +760,97 @@ def _get_images_dir():
...
+            two_zvols:
+              - system:
+                  image: zroot/bhyve/CentOS-7-x86_64-v1@v1.0.5
+                  hostname_property: virt:hostname
+                  disk_filesystem: zroot/bhyve/guests
+              - data:
+                  disk_filesystem: tank/disks
+                  size: 20G
+                  hostname_property: virt:hostname
+                  sparse_volume: True
```
Comment (cbosdo) [inline salt/modules/virt.py]: This documentation will not be visible to the end user: move this into the ``virt.init`` disks documentation

File: salt/modules/virt.py [hunk 8]
```diff
@@ -892,6 +993,10 @@ def _disk_profile(profile, hypervisor, disks=None, vm_name=None, image=None, poo
     elif hypervisor in ['qemu', 'kvm']:
         overlay = {'format': 'qcow2',
                    'model': 'virtio'}
+    elif hypervisor in ['bhyve']:
+        overlay = {'format': 'raw',
+                   'model': 'virtio',
+                   'sparse_volume': False}
```
Comment (cbosdo) [inline salt/modules/virt.py]: you'll need to change the function name in a separate commit since it doesn't create only qemu images ;)

File: salt/modules/virt.py [hunk 9]
```diff
@@ -2801,14 +2921,26 @@ def purge(vm_, dirs=False, removables=None, **kwargs):
     for disk in disks:
         if not removables and disks[disk]['type'] in ['cdrom', 'floppy']:
             continue
-        os.remove(disks[disk]['file'])
-        directories.add(os.path.dirname(disks[disk]['file']))
+        elif disks[disk].get('zfs', False):
+            # Sleep a little while to avoid 'dataset is busy'
+            time.sleep(3)
```
Comment (cbosdo) [inline salt/modules/virt.py]: That is fairly dangerous and can strike back. Is there no reliable way to know when the disk can be safely removed?

File: salt/modules/virt.py [hunk 10]
```diff
@@ -2801,14 +2921,26 @@ def purge(vm_, dirs=False, removables=None, **kwargs):
...
     if dirs:
         for dir_ in directories:
             shutil.rmtree(dir_)
     if getattr(libvirt, 'VIR_DOMAIN_UNDEFINE_NVRAM', False):
         # This one is only in 1.2.8+
-        dom.undefineFlags(libvirt.VIR_DOMAIN_UNDEFINE_NVRAM)
+        try:
+            dom.undefineFlags(libvirt.VIR_DOMAIN_UNDEFINE_NVRAM)
+        except libvirt.libvirtError as ex:
+            dom.undefine()
```
Comment (cbosdo) [inline salt/modules/virt.py]: That hunk should go in a separate commit since that has nothing to do with bhyve

File: salt/modules/virt.py [hunk 11]
```diff
@@ -3466,6 +3610,7 @@ def _parse_caps_guest(guest):
             'machines': {},
             'domains': {}
         },
+        'features': {}
```
Comment (cbosdo) [inline salt/modules/virt.py]: This will likely break the capabilities unit tests, and I'm not even sure it is needed with the next hunk

File: salt/modules/virt.py [hunk 12]
```diff
@@ -3664,7 +3809,6 @@ def capabilities(**kwargs):
     conn = __get_conn(**kwargs)
     caps = _capabilities(conn)
     conn.close()
-
```
Comment (cbosdo) [inline salt/modules/virt.py]: Keep the line around. Cosmetic changes should go in their own commit.

File: salt/templates/virt/libvirt_domain.jinja [hunk 13]
```diff
@@ -7,6 +7,9 @@
                 {% for dev in boot_dev %}
                 <boot dev='{{ dev }}' />
                 {% endfor %}
+		{% if loader %}
+		<loader readonly='yes' type='pflash'>{{ loader }}</loader>
```
Comment (cbosdo) [inline salt/templates/virt/libvirt_domain.jinja]: ``readonly`` should be configurable and allow a ``no`` value.
``type`` should be configurable too to allow the ``rom`` value
While at it, could you add a ``secure`` flag (value yes or no) as documented in https://libvirt.org/formatdomain.html#elementsOS
Add the whole loader section in the template should go in it own commit since that can also be used for Xen or QEMU guests.

File: salt/templates/virt/libvirt_domain.jinja [hunk 14]
```diff
@@ -51,7 +54,11 @@
                             {% endif %}/>
                 </graphics>
                 {% endif %}
-                {% if serial_type == 'pty' %}
+		{% if serial_type == 'nmdm' %}
+		<serial type='nmdm'>
+			<source master='/dev/nmdm{{ nmdm_id }}A' slave='/dev/nmdm{{ nmdm_id }}B'/>
+		</serial>
+                {% elif serial_type == 'pty' %}
```
Comment (cbosdo) [inline salt/templates/virt/libvirt_domain.jinja]: The indentation is off here. Please adjust to match the surrounding code.

Comment (cbosdo) [reply salt/templates/virt/libvirt_domain.jinja]: I'ld love to see some unit tests for the new serial type and the loader element to avoid breaking the features you need with future potential refactoring.

**Review by cbosdo — CHANGES_REQUESTED** [review]
Could you please address the comments and resubmit?

**Comment by cbosdo** [comment]
Note, there may be a race with PR #48736, the last one to be in needs to rebase on the other ;)

File: salt/modules/virt.py (hunk 7 above)
Comment (jeroen92) [reply salt/modules/virt.py]: That goes for a lot of documented functions (i.e. disk_profile). The dict keys should be documented at init(), but I feel the rest of the docs are a bit out of place there..

File: salt/modules/virt.py (hunk 3 above)
Comment (jeroen92) [reply salt/modules/virt.py]: Now bhyve has good support for VNC I'd suggest to omit the nmdm serial type.

File: salt/modules/virt.py (hunk 6 above)
Comment (jeroen92) [reply salt/modules/virt.py]: Check, fixed this.

File: salt/modules/virt.py (hunk 5 above)
Comment (jeroen92) [reply salt/modules/virt.py]: Ack.

File: salt/modules/virt.py (hunk 8 above)
Comment (jeroen92) [reply salt/modules/virt.py]: Change which function name?

File: salt/modules/virt.py (hunk 9 above)
Comment (jeroen92) [reply salt/modules/virt.py]: After destroying a domain the OS seems to keep the volume openened for some time. I'm not sure yet how to create a better workaround for this other than waiting a bit.

File: salt/modules/virt.py (hunk 10 above)
Comment (jeroen92) [reply salt/modules/virt.py]: It does. On FreeBSD calling undefineFlags raises `libvirt: Domain Config error : this function is not supported by the connection driver: virDomainUndefineFlags`

Comment (cbosdo) [reply salt/modules/virt.py]: ok, but I suspect this is not the only driver in that case. I still think it would be better to have it in a separate commit with what you just wrote as commit description to explain the reason of the change

File: salt/modules/virt.py (hunk 8 above)
Comment (jeroen92) [reply salt/modules/virt.py]: I didn't touch the contents of _qemu_image_create(). The code block referenced by your comment resides in _disk_profile()

File: salt/modules/virt.py (hunk 9 above)
Comment (cbosdo) [reply salt/modules/virt.py]: OK, then let's do it this way until you have a better idea

File: salt/modules/virt.py (hunk 8 above)
Comment (cbosdo) [reply salt/modules/virt.py]: oh my bad... sorry for the inconvenience, it looks like I misread the patch ;)

File: salt/templates/virt/libvirt_domain.jinja (hunk 13 above)
Comment (jeroen92) [reply salt/templates/virt/libvirt_domain.jinja]: Ack, changed the loader argument accordingly.

**Comment by jeroen92** [comment]
I haven't tested the unittests yet, since I was getting all kinds of unicode related errors. So those are still pending, as well as splitting the commits.

`   .............................................................................
   -> unit.modules.test_virt.VirtTestCase.test_gen_xml_loader_default  .........
       Traceback (most recent call last):
         File "/root/salt/tests/unit/modules/test_virt.py", line 361, in test_gen_xml_loader_default
           self.assertEqual(root.find('os/loader').attrib['type'], 'pflash')
       AssertionError: 'fplash' != u'pflash'
   .............................................................................
`

**Comment by cbosdo** [comment]
@jeroen92 How are you running the tests? the assert looks good to me. Which version of python is that?

**Comment by jeroen92** [comment]
See below on how I'm running the tests. I've followed `Salt's Test Suite: An Introduction` when setting up the develop environment. Besides the unicode thingie, I'm also getting a bunch of other errors though.

```
$ python runtests.py --name unit.modules.test_virt
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 * Python Version: 2.7.15 (default, Jun 6 2018, 21:13:14) [GCC 4.2.1 Compatible FreeBSD Clang 4.0.0 (tags/RELEASE_400/final 297347)]
 * Transplanting configuration files to '/tmp/salt-tests-tmpdir/config'
 * Current Directory: /root/salt
 * Test suite is running under PID 89553
 * Logging tests on /tmp/salt-runtests.log
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Starting unit.modules.test_virt Tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
EEEFEE..........EEEEEEEEEFEEEEEEE...................EE
======================================================================
ERROR: test_boot_custom_dev (unit.modules.test_virt.VirtTestCase)
[CPU:44.5%|MEM:97.3%]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/salt/tests/unit/modules/test_virt.py", line 136, in test_boot_custom_dev
    boot_dev='cdrom'
  File "/root/salt/salt/modules/virt.py", line 627, in _gen_xml
    for key, val in loader.items():
AttributeError: 'NoneType' object has no attribute 'items'

======================================================================
ERROR: test_boot_default_dev (unit.modules.test_virt.VirtTestCase)
[CPU:50.0%|MEM:97.3%]
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/salt/tests/unit/modules/test_virt.py", line 118, in test_boot_default_dev
    'kvm'
  File "/root/salt/salt/modules/virt.py", line 627, in _gen_xml
    for key, val in loader.items():
AttributeError: 'NoneType' object has no attribute 'items'
....
```

**Comment by cbosdo** [comment]
@jeroen92 What I do is that I created a virtualenv in the base folder of the tree, I source it's ``activate`` file, then installed all the py3 deps. to run the test I just

    cd tests
    ./runtests.py -n unit.modules.test_virt

**Comment by jeroen92** [comment]
@cbosdo Updated the MR. All tests pass and discussions are resolved.

**Comment by cachedout** [comment]
Bump @cbosdo for re-review please. :)

File: salt/modules/virt.py [hunk 15]
```diff
@@ -454,6 +468,11 @@ def _get_disks(dom):
             qemu_target = source.get('file', '')
             if not qemu_target:
                 qemu_target = source.get('dev', '')
+            if source.get('file').startswith('/dev/zvol/'):
```
Comment (cbosdo) [inline salt/modules/virt.py]: The problem with ``source.get('file')`` is that you may get ``None`` if no there is no ``file``. Either you move that block above the ``if not qemu_target:`` and use ``qemu_target`` value or you need to default to the empty string.

File: salt/modules/virt.py [hunk 16]
```diff
@@ -2095,6 +2258,33 @@ def get_graphics(vm_, **kwargs):
...
 
 
+def get_loader(vm_, **kwargs):
+    '''
+    Returns the information on the loader for a given vm
+
+    :param vm_: name of the domain
+    :param connection: libvirt connection URI, overriding defaults
+
+        .. versionadded:: Fluorine
```
Comment (cbosdo) [inline salt/modules/virt.py]: You don't need to have a ``.. versionadded:: Fluorine`` on all these properties since the whole function is brand new. Just add one ``.. versionadded`` on the function level.

**Review by cbosdo — CHANGES_REQUESTED** [review]
Sorry for the late review: I'm just back from vacation.

Looks good to me, with two small things to fix. Next time we'll be able to push ;)

**Comment by cachedout** [comment]
@cbosdo Thank you! (Hope your vacation was wonderful.)

@jeroen92 Could you review the latest feedback?

**Comment by jeroen92** [comment]
@cachedout @cbosdo Updated :)

**Review by cbosdo — APPROVED** [review]
LGTM, thanks for the effort!

**Review by rallytime — APPROVED** [review]
LGTM

**Comment by rallytime** [comment]
@jeroen92 There is a related test failing on this change. Can you take a look?

https://jenkinsci.saltstack.com/job/pr-kitchen-centos7-py3/job/PR-47620/11/

**Comment by jeroen92** [comment]
@rallytime Thanks for checking, I missed that one. Should be fixed now.
